python/tempo/tsdf.py

```diff
@@ -1,533 +1,623 @@
+        :param prefix - supply a prefix for the newly sampled columns
+        :param fill - Boolean - set to True if the desired output should contain filled in gaps (with 0s currently)
+        :return: TSDF object with sample data using aggregate function
+        """
+        rs.validateFuncExists(func)
+        enriched_tsdf = rs.aggregate(self, freq, func, metricCols, prefix, fill)
+        return (enriched_tsdf)
+
+    def calc_bars(tsdf, freq, func=None, metricCols=None, fill=None):
+
+        resample_open = tsdf.resample(freq=freq, func='floor', metricCols=metricCols, prefix='open', fill=fill)
+        resample_low = tsdf.resample(freq=freq, func='min', metricCols=metricCols, prefix='low', fill=fill)
+        resample_high = tsdf.resample(freq=freq, func='max', metricCols=metricCols, prefix='high', fill=fill)
+        resample_close = tsdf.resample(freq=freq, func='ceil', metricCols=metricCols, prefix='close', fill=fill)
+
+        join_cols = resample_open.partitionCols + [resample_open.ts_col]
+        bars = resample_open.df.join(resample_high.df, join_cols).join(resample_low.df, join_cols).join(
+            resample_close.df, join_cols)
+        non_part_cols = set(set(bars.columns) - set(resample_open.partitionCols)) - set([resample_open.ts_col])
+        sel_and_sort = resample_open.partitionCols + [resample_open.ts_col] + sorted(non_part_cols)
+        bars = bars.select(sel_and_sort)
+
+        return (TSDF(bars, resample_open.ts_col, resample_open.partitionCols))
+
+    def fourier_transform(self, timestep, valueCol):
+        """
+        Function to fourier transform the time series to its frequency domain representation.
+        :param timestep: timestep value to be used for getting the frequency scale
+        :param valueCol: name of the time domain data column which will be transformed
+        """
+
+        def tempo_fourier_util(pdf):
+            """
+            This method is a vanilla python logic implementing fourier transform on a numpy array using the scipy module.
+            This method is meant to be called from Tempo TSDF as a pandas function API on Spark
+            """
+            # global TIMESTEP
+            select_cols = list(pdf.columns)
+            y = np.array(pdf['tdval'])
+            tran = fft(y)
+            r = tran.real
+            i = tran.imag
+            pdf['ft_real'] = r
+            pdf['ft_imag'] = i
+            N = tran.shape
+            # timestep = TIMESTEP
+            xf = fftfreq(N[0], timestep)
+            pdf['freq'] = xf
+            return pdf[select_cols + ['freq', 'ft_real', 'ft_imag']]
+
+        valueCol = self.__validated_column(self.df, valueCol)
+        # self.set_timestep(timestep)
+        if self.sequence_col:
+            data = self.df.orderBy(self.ts_col, self.sequence_col)
+            if self.partitionCols == []:
```

Comment:
@Spratiher9 , is this orderBy necessary? Any orderBy or sort is going to incur a big penalty. Instead please sort inside the pandas UDF